include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -3353,6 +3353,9 @@ struct OrtCustomOp {
   // Returns the characteristics of the input & output tensors
   OrtCustomOpInputOutputCharacteristic(ORT_API_CALL* GetInputCharacteristic)(_In_ const struct OrtCustomOp* op, _In_ size_t index);
   OrtCustomOpInputOutputCharacteristic(ORT_API_CALL* GetOutputCharacteristic)(_In_ const struct OrtCustomOp* op, _In_ size_t index);
+
+  // Returns the memory type of the input tensors
+  OrtMemType(ORT_API_CALL* GetInputMemoryType)(_In_ const struct OrtCustomOp* op, _In_ size_t index);
```

Comment:
Nit: Might as well add a similar interface for outputs as well ? (Although it will have to be a very rare kind of operator to output something on CPU while executing on CUDA)

Reply:
Unless you have an explicit use case in mind I would hold off on doing so as you're probably adding something that will never be used. I can't think of any non-CPU op (apart from the copy to host ones) that would produce an output on a different device. Typically we only use the CPU input for values that have to be used in CPU based logic (e.g. iteration count for a loop).